# Curso de Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la version que esta usando**

In [9]:
import pandas as pd


In [8]:
print(pd.__version__)

3.0.2


**leer el archivo Salaries.csv en un dataframe llamado salarios**

se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [10]:
# Copie el código aca
import pandas as pd
url = "https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv"
salarios = pd.read_csv(url)
salarios.head()


,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [11]:
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  str    
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), str(3)
memory usage: 14.7 MB


- Segun la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuantas filas y columnas tiene luego de eliminar las columnas**

In [12]:
# 1. Eliminamos las columnas 'Notes' y 'Status' de forma permanente
salarios.drop(['Notes', 'Status'], axis=1, inplace=True)

# 2. Visualizamos cuántas filas y columnas quedaron usando .shape
print("El tamaño actual de la tabla (filas, columnas) es:")
print(salarios.shape)

El tamaño actual de la tabla (filas, columnas) es:
(148654, 11)


**¿Cual es la nueva información del dataframe?**

In [13]:
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame.**

In [14]:
salarios.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los ultimos 5 registros del DataFrame.**

In [15]:
salarios.tail()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuantas filas y columnas tiene el dataframe?**

In [18]:
filas = salarios.shape[0]
columnas = salarios.shape[1]

print(f"El DataFrame tiene {filas} filas y {columnas} columnas.")

El DataFrame tiene 148654 filas y 11 columnas.


**Observe los 10 registros que estan desde el indice 3000**

In [19]:
salarios.iloc[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco



La columna *Benefits* tiene valores `NaN`, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [20]:
# 1. Calculamos la cantidad de valores NaN en la columna 'Benefits'
cantidad_nan = salarios['Benefits'].isna().sum()

print(f"La columna 'Benefits' tiene {cantidad_nan} valores nulos (NaN).")

La columna 'Benefits' tiene 36163 valores nulos (NaN).


Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [22]:
# 1. Rellenamos los NaN de la columna 'Benefits' con el número 0 de forma permanente
salarios.fillna({'Benefits': 0}, inplace=True)

# 2. Mostramos los primeros 10 elementos para comprobar el cambio
salarios.head(10)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,0.0,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,0.0,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,0.0,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,0.0,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,0.0,326373.19,326373.19,2011,San Francisco
5,6,DAVID SULLIVAN,ASSISTANT DEPUTY CHIEF II,118602.00,8601.00,189082.74,0.0,316285.74,316285.74,2011,San Francisco
6,7,ALSON LEE,"BATTALION CHIEF, (FIRE DEPARTMENT)",92492.01,89062.90,134426.14,0.0,315981.05,315981.05,2011,San Francisco
7,8,DAVID KUSHNER,DEPUTY DIRECTOR OF INVESTMENTS,256576.96,0.00,51322.50,0.0,307899.46,307899.46,2011,San Francisco
8,9,MICHAEL MORRIS,"BATTALION CHIEF, (FIRE DEPARTMENT)",176932.64,86362.68,40132.23,0.0,303427.55,303427.55,2011,San Francisco
9,10,JOANNE HAYES-WHITE,"CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)",285262.00,0.00,17115.73,0.0,302377.73,302377.73,2011,San Francisco


**¿Cual es la nueva información del dataframe?**

In [23]:
salarios.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


**Cuantos valores unicos hay en las columnas categoricas?**

In [24]:
# 1. Seleccionamos solo las columnas categóricas (las que son de tipo texto 'object')
columnas_categoricas = salarios.select_dtypes(include=['object'])

# 2. Contamos cuántos valores únicos tiene cada una de esas columnas
valores_unicos = columnas_categoricas.nunique()

# 3. Imprimimos el resultado
print(valores_unicos)

EmployeeName    110811
JobTitle          2159
Agency               1
dtype: int64


C:\Users\Yesenia Rivera\AppData\Local\Temp\ipykernel_1116\1935884082.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  columnas_categoricas = salarios.select_dtypes(include=['object'])


- la columna **Agency** solamente tiene un valor unico, entonces no aporta informacion, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [25]:
# Eliminamos la columna 'Agency' de forma permanente
salarios.drop('Agency', axis=1, inplace=True)

print("Columnas restantes en el DataFrame:")
print(salarios.columns.tolist())

Columnas restantes en el DataFrame:
['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year']


**Cual es el promedio de BasePay ?**

In [26]:
# Calculamos el promedio de la columna 'BasePay'
promedio_basepay = salarios['BasePay'].mean()

print(f"El promedio del salario base (BasePay) es: {promedio_basepay}")

El promedio del salario base (BasePay) es: 66325.44884048769


**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [27]:
# Calculamos el valor máximo de la columna 'OvertimePay'
maximo_overtime = salarios['OvertimePay'].max()

print(f"La cantidad más alta de pago por horas extras (OvertimePay) es: {maximo_overtime}")

La cantidad más alta de pago por horas extras (OvertimePay) es: 245131.88


**Realice una descripcion estadistica de las columnas numericas?**

In [28]:
salarios.describe()

,Id,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
count,148654.000000,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000,148654.000000
mean,74327.500000,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811,2012.522643
std,42912.857795,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483,1.117538
min,1.000000,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000,2011.000000
25%,37164.250000,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000,2012.000000
50%,74327.500000,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000,2013.000000
75%,111490.750000,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000,2014.000000
max,148654.000000,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000,2014.000000


**En este dataset las variables `Id` y `Year` representan variables categorica, convertir estas columnas a su formato correcto**

In [29]:
# Convertimos 'Id' y 'Year' al formato categórico
salarios['Id'] = salarios['Id'].astype('category')
salarios['Year'] = salarios['Year'].astype('category')

print(salarios[['Id', 'Year']].dtypes)

Id      category
Year    category
dtype: object


**la variable `Year` es categorica ordinal, primero detecte cuales son los valores unicos que tiene la columna**

In [30]:
años_unicos = salarios['Year'].unique()

print(f"Los años registrados en esta base de datos son: {años_unicos}")

Los años registrados en esta base de datos son: [2011, 2012, 2013, 2014]
Categories (4, int64): [2011, 2012, 2013, 2014]


**Apartir de la información anterior convierta la columna `Year` en variable categorica ordinal y visualice la columna**

In [31]:
from pandas.api.types import CategoricalDtype

# Tomamos los años únicos y los ordenamos de menor a mayor
años_ordenados = sorted(salarios['Year'].unique())

# Creamos la "regla" de que es una categoría ordinal y le pasamos el orden
tipo_ordinal = CategoricalDtype(categories=años_ordenados, ordered=True)

# Le aplicamos esta nueva regla a la columna 'Year'
salarios['Year'] = salarios['Year'].astype(tipo_ordinal)

# Visualizamos la columna
salarios['Year']

0         2011
1         2011
2         2011
3         2011
4         2011
          ... 
148649    2014
148650    2014
148651    2014
148652    2014
148653    2014
Name: Year, Length: 148654, dtype: category
Categories (4, int64): [2011 < 2012 < 2013 < 2014]

**Realice nuevamente una descripcion estadistica de las columnas numericas?**

In [32]:
salarios.describe()

,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits
count,148045.000000,148650.000000,148650.000000,148654.000000,148654.000000,148654.000000
mean,66325.448840,5066.059886,3648.767297,18924.232839,74768.321972,93692.554811
std,42764.635495,11454.380559,8056.601866,17165.279495,50517.005274,62793.533483
min,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000
25%,33588.200000,0.000000,0.000000,0.000000,36168.995000,44065.650000
50%,65007.450000,0.000000,811.270000,23214.020000,71426.610000,92404.090000
75%,94691.050000,4658.175000,4236.065000,33468.980000,105839.135000,132876.450000
max,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000


**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [33]:
puesto_joseph = salarios[salarios['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

print(puesto_joseph)

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: str


**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [34]:
pago_joseph = salarios[salarios['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits'].iloc[0]

print(f"JOSEPH DRISCOLL gana un total de: ${pago_joseph}")

JOSEPH DRISCOLL gana un total de: $270324.91


**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [38]:
# Buscamos la fila completa del empleado con el pago más alto
salarios[salarios['TotalPayBenefits'] == salarios['TotalPayBenefits'].max()]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.0,400184.25,0.0,567595.43,567595.43,2011


**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [39]:
# Buscamos la fila completa del empleado con el pago más bajo (incluyendo beneficios)
salarios[salarios['TotalPayBenefits'] == salarios['TotalPayBenefits'].min()]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014


*Interpretación:**
Sí, efectivamente el valor de su salario está en negativo. 

**Cuantos registros tienen un valor de TotalPayBeneficts menor que cero**

In [40]:
# Creamos una máscara booleana para los valores menores a cero
cantidad_negativos = (salarios['TotalPayBenefits'] < 0).sum()

# Imprimimos el resultado organizado
print(f"Hay {cantidad_negativos} registros con un valor de TotalPayBenefits menor que cero.")

Hay 4 registros con un valor de TotalPayBenefits menor que cero.


**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [41]:
# Agrupamos por la columna 'Year'
# Seleccionamos la columna 'BasePay'
# Calculamos el promedio (.mean())
promedio_anual = salarios.groupby('Year')['BasePay'].mean()

# 4. Mostramos el resultado
print(promedio_anual)

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64


**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [42]:
# Contamos la cantidad de títulos de trabajo únicos
total_titulos = salarios['JobTitle'].nunique()

print(f"Existen {total_titulos} títulos de trabajo únicos en el dataset.")

Existen 2159 títulos de trabajo únicos en el dataset.


**¿Cuales son los 5 trabajos mas comunes?**

In [43]:
# Contamos las repeticiones de cada JobTitle
top_5_trabajos = salarios['JobTitle'].value_counts().head(5)

print(top_5_trabajos)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64


**¿Cuántos puestos de trabajo Unicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [44]:
# Filtramos el DataFrame para quedarnos solo con el año 2013
salarios_2013 = salarios[salarios['Year'] == 2013]

# Contamos cuántas veces aparece cada JobTitle en ese año específico
conteo_puestos = salarios_2013['JobTitle'].value_counts()

# Sumamos cuántos de esos puestos tienen un conteo exactamente igual a 1
puestos_unicos_2013 = (conteo_puestos == 1).sum()

print(f"En 2013, hubo {puestos_unicos_2013} títulos de trabajo que solo ocupaba una persona.")

En 2013, hubo 202 títulos de trabajo que solo ocupaba una persona.


**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [45]:
# Definimos la función de búsqueda
def tiene_chief(titulo):
    # Convertimos a minúsculas para que no importe cómo esté escrito
    if 'chief' in titulo.lower():
        return True
    else:
        return False

# Aplicamos la función a la columna y sumamos los True
cantidad_chiefs = salarios['JobTitle'].apply(tiene_chief).sum()

# Imprimimos el resultado
print(f"Hay {cantidad_chiefs} personas que tienen la palabra 'Chief' en su título de trabajo.")

Hay 627 personas que tienen la palabra 'Chief' en su título de trabajo.


Este código es valido, se recomienda hacer el código mas simple y agregar typing y documentación.

In [49]:
def buscar_palabra_en_puestos(df: pd.DataFrame, palabra: str) -> int:
    """
    Cuenta cuántos registros en la columna 'JobTitle' contienen una palabra específica.
    
    Args:
        df (pd.DataFrame): El DataFrame que contiene los datos de salarios.
        palabra (str): La palabra clave a buscar (ej. 'chief').
        
    Returns:
        int: El número total de personas que tienen esa palabra en su cargo.
    """
    # Usamos str.contains con case=False para ignorar mayúsculas/minúsculas
    # y na=False para evitar errores si hay celdas vacías.
    conteo = df['JobTitle'].str.contains(palabra, case=False, na=False).sum()
    return int(conteo)

# Ejecutamos la función y guardamos el resultado
cantidad_chiefs = buscar_palabra_en_puestos(salarios, 'chief')

# Mostramos el resultado final
print(f"Resultado: {cantidad_chiefs} personas tienen la palabra 'Chief' en su título de trabajo.")

Resultado: 627 personas tienen la palabra 'Chief' en su título de trabajo.


**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

crea una nueva columna donde se guarde la longitud del titulo del trabajo. tip: usar el método `.apply()`

In [53]:
# Creamos la columna con la longitud
salarios['title_len'] = salarios['JobTitle'].apply(len)

# Llamamos a la columna para que se imprima el resultado
salarios['title_len']

0         46
1         31
2         31
3         36
4         44
          ..
148649     9
148650    12
148651    12
148652    12
148653    26
Name: title_len, Length: 148654, dtype: int64

Utilice el método `.corr()` de Pandas para calcular la correlacion

In [54]:
salarios['title_len'] = salarios['JobTitle'].apply(len)

salarios[['title_len', 'TotalPayBenefits']].corr()

,title_len,TotalPayBenefits
title_len,1.000000,-0.036878
TotalPayBenefits,-0.036878,1.000000


*Respuesta:* No existe una correlación. El largo del nombre del puesto no determina el salario, ya que el número obtenido es casi cero